tests/testthat/test_utils.r

```diff
@@ -13,10 +11,10 @@ test_that('skip_repeated does not skip three or less consecutive items', {
 })
 
 test_that('skip_repeated works on tracebacks', {
-    err <- try_capture_stack(quote({
+    err <- evaluate::try_capture_stack(quote({
         f <- function(x) stop(x)
         f(1)
     }), new.env())
     skipped_stack <- skip_repeated(err$calls)
-    expect_is(skipped_stack[[1]], 'call')
+    expect_type(skipped_stack[[1]], 'language')
```

Comment:
`expect_is()` is deprecated from testthat3e:

https://testthat.r-lib.org/reference/expect_is.html

Note that `typeof(call('sum'))` is `"language"`; also "call" is the C type `LANGSXP`. You can see that `is.call(x)` is actually just checking `TYPEOF(x) == LANGSXP` here:

https://github.com/r-devel/r-svn/blob/11aec6f992e555f59c144b0780ff314e12f4097b/src/main/coerce.c#L2129-L2131

---

I can either (1) Use `expect_true(is.call(x))` or (2) add a comment similar to the above like `expect_type(., "language") # i.e., this is a call`

WDYT?

Reply:
Note further:

`typeof(x) == "language"` and `is.language(x)` have (quite confusingly!!) different behavior. This is noted in `?language`:

> If `typeof(x) == "language"`, then `is.language(x)` is always true, but the reverse does not hold as expressions or names `y` also fulfill `is.language(y)`

Reply:
Not quite, my point is that this isn’t true: `expect_type(., "language")  # i.e., this is a call`

See e.g. the docs of `is.call`:

> `is.language()` is also true for `call`s (but also for `symbol`s and `expression`s where `is.call()` is false). 

Or the next `case` in the C code you linked. `language` is broader, `call` is narrower.

I want the test to keep testing the narrower case, so `expect_true(is.call(x))` would be one way to do it.

Reply:
I still don't think we're on the same page, and I still think `is.call()` is correct. Here's all the `SEXPTYPE`s:

https://github.com/r-devel/r-svn/blob/601c09a712de2ecd58d3dd976b455a6721685afb/src/include/Rinternals.h#L144-L174

I don't know how to build all of them from R, but here's everything that I can build directly from R:

```r
types <- list(
  NILSXP = NULL,
  SYMSXP = as.name('a'),
  # LISTSXP = ?
  CLOSXP = lm,
  ENVSXP = .GlobalEnv,
  # PROMSXP = ?
  LANGSXP = call("a"),
  SPECIALSXP = `$`,
  BUILTINSXP = `+`,
  # CHARSXP = <not accessible from R>
  LGLSXP = TRUE,
  INTSXP = 1L,
  REALSXP = 1.0,
  CPLXSXP = complex(1.0),
  STRSXP = "a",
  # DOTSXP = ?
  # ANYSXP = ?
  VECSXP = list(),
  EXPRSXP = expression(),
  # BCODESXP = ?
  EXTPTRSXP = xml2::xml_new_document()$doc,
  # WEAKREFSXP = ?
  RAWSXP = as.raw(1),
  OBJSXP = Matrix::sparseMatrix(1, 1)
  # NEWSXP = ?
  # FREESXP = ?
  # FUNSXP = ?
)
```

AIUI this is an exhaustive list of what you'll encounter from R.

Now let's compare `is.call()`, `class()`, `is.language()`, `typeof()`:

```r
data.frame(
  is.call = sapply(types, is.call),
  class = sapply(types, \(x) toString(class(x))),
  is.language = sapply(types, is.language),
  typeof = sapply(types, typeof)
)
```

With output:

```
           is.call       class is.language      typeof
NILSXP       FALSE        NULL       FALSE        NULL
SYMSXP       FALSE        name        TRUE      symbol
CLOSXP       FALSE    function       FALSE     closure
ENVSXP       FALSE environment       FALSE environment
LANGSXP       TRUE        call        TRUE    language
SPECIALSXP   FALSE    function       FALSE     special
BUILTINSXP   FALSE    function       FALSE     builtin
LGLSXP       FALSE     logical       FALSE     logical
INTSXP       FALSE     integer       FALSE     integer
REALSXP      FALSE     numeric       FALSE      double
CPLXSXP      FALSE     complex       FALSE     complex
STRSXP       FALSE   character       FALSE   character
VECSXP       FALSE        list       FALSE        list
EXPRSXP      FALSE  expression        TRUE  expression
EXTPTRSXP    FALSE externalptr       FALSE externalptr
RAWSXP       FALSE         raw       FALSE         raw
OBJSXP       FALSE   ngCMatrix       FALSE          S4
```

i.e., `is.call()` and `typeof(.)=="language"` correspond exactly to `LANGSXP` and `call('a')`.

Reply:
More specifically to the interaction with testthat, we can add more columns:

```r
library(testthat)
pass_test <- \(...) !inherits(tryCatch(..., error = identity), "error")
data.frame(
  expect_is_call = sapply(types, \(x) pass_test(expect_is(x, "call"))),
  expect_type_language = sapply(types, \(x) pass_test(expect_type(x, "language"))),
  expect_true_is.call = sapply(types, \(x) pass_test(expect_true(is.call(x))))
)
```

With results:

```
           expect_is_call expect_type_language expect_true_is.call
NILSXP              FALSE                FALSE               FALSE
SYMSXP              FALSE                FALSE               FALSE
CLOSXP              FALSE                FALSE               FALSE
ENVSXP              FALSE                FALSE               FALSE
LANGSXP              TRUE                 TRUE                TRUE
SPECIALSXP          FALSE                FALSE               FALSE
BUILTINSXP          FALSE                FALSE               FALSE
LGLSXP              FALSE                FALSE               FALSE
INTSXP              FALSE                FALSE               FALSE
REALSXP             FALSE                FALSE               FALSE
CPLXSXP             FALSE                FALSE               FALSE
STRSXP              FALSE                FALSE               FALSE
VECSXP              FALSE                FALSE               FALSE
EXPRSXP             FALSE                FALSE               FALSE
EXTPTRSXP           FALSE                FALSE               FALSE
RAWSXP              FALSE                FALSE               FALSE
OBJSXP              FALSE                FALSE               FALSE
```

Reply:
@hadley did point out one edge case neither of us caught on to, though:

https://github.com/r-lib/testthat/issues/1448#issuecomment-1810177605

`class(quote({}))` is `"{"` while `typeof(quote({}))` is `"language"` and `is.call(quote({}))` is `TRUE`. So there's a sliver of difference between `expect_is(x, "call")` and `expect_true(is.call(x))`. This is all coming from the C routine `lang2str` which gives `class(x)` for `LANGSXP`:

https://github.com/r-devel/r-svn/blob/423f5290506d31263c6b91b272beb14df97a278a/src/main/attrib.c#L612-L635

In total there are 7 such cases where we can have a `LANGSXP` in hand but `class(x)` is something other than `"call"`.

If that is important for the current test, I think `expect_true(inherits(x, "call"))` is the "proper" approach.

Reply:
Maybe that’s the best. It’s been a long time since I touched those tests. I think the test is supposed to check if the object is equivalent to `quote(f(1))`, i.e. a function call.

So maybe `expect_identical(mode(x), "call")`?

Reply:
Unfortunately that's also not what we want -- `mode(quote({}))` is also `"call"`.

Reply:
OK, PTAL. Thanks for the discussion!